mlrun/projects/pipelines.py

```diff
@@ -209,6 +211,57 @@ def is_initialized(self, raise_exception=False):
 pipeline_context = _PipelineContext()
 
 
+def _set_priority_class_name_on_kfp_pod(kfp_pod_template, function):
+    if kfp_pod_template.get("container") and kfp_pod_template.get("name").startswith(
+        function.metadata.name
+    ):
+        kfp_pod_template["PriorityClassName"] = getattr(
+            function.spec, "priority_class_name", ""
+        )
+
+
+# When we run pipelines, the kfp.compile.Compile.compile() method takes the decorated function with @dsl.pipeline and
+# converts it to a k8s object. As part of the flow in the Compile.compile() method,
+# we call _create_and_write_workflow, which builds a dictionary from the workflow and then writes it to a file.
+# Unfortunately, the kfp sdk does not provide an API for configuring priority_class_name and other attributes.
+# I ran across the following problem when seeking for a method to set the priority_class_name:
+# https://github.com/kubeflow/pipelines/issues/3594
+# When we patch the _create_and_write_workflow, we can eventually obtain the dictionary right before we write it
+# to a file and enrich it with argo compatible fields, make sure you looking for the same argo version we use
+# https://github.com/argoproj/argo-workflows/blob/release-2.7/pkg/apis/workflow/v1alpha1/workflow_types.go
+def _create_and_write_workflow(
+    self,
+    pipeline_func,
+    pipeline_name=None,
+    pipeline_description=None,
+    params_list=None,
+    pipeline_conf=None,
+    package_path=None,
+) -> None:
+    """Compile the given pipeline function and dump it to specified file
+    format."""
+    workflow = self._create_workflow(
+        pipeline_func, pipeline_name, pipeline_description, params_list, pipeline_conf
+    )
+    # enrich each pipeline step with your desire k8s attribute
+    for kfp_pod_template in workflow["spec"]["templates"]:
+        if kfp_pod_template.get("container"):
+            if pipeline_context.functions:
+                for (
+                    function_name,
+                    function_obj,
+                ) in pipeline_context.functions.items().items():
+                    _set_priority_class_name_on_kfp_pod(kfp_pod_template, function_obj)
+                    break
+
+    self._write_workflow(workflow, package_path)
+    kfp.compiler.compiler._validate_workflow(workflow)
+
+
+# patching function as class method
+kfp.compiler.Compiler._create_and_write_workflow = _create_and_write_workflow
```

Comment:
Super duper hacky, when doing such things, we must have tests protecting the hack from breaking, I suggest:
1. Have a unit test that is just doing the compile and verifying the output yaml has the priority class we want - this should protect us from a scenario like us upgrading the kfp package version, which changes the method name, which will cause this hack not to work.
2. Have a system test that is really running a pipeline, listing the pods verifying that the right priority class was assigned to them - This will protect us from a scenario like us upgrading the kfp backend (pipelines service in Iguazio) which stops support this for some reason